In [1]:
# utilizando la libreria os para manejos de directorios
import os

# Computacion vectorial y cientifica para python
import numpy as np

#importamos pandas para el manejo del dataset, y separarlos dentro de una matriz
import pandas as pd

#esta tabulate nos sirve para hacer tablas
from tabulate import tabulate

# Librerias para graficación (trazado de gráficos)
from matplotlib import pyplot
from mpl_toolkits.mplot3d import Axes3D  # -> Necesario para graficar superficies 3D

#Para separa el 20% y 80%
from sklearn.model_selection import train_test_split

# Modulo de optimizacion en scipy
from scipy import optimize

In [ ]:
from google.colab import drive
drive.mount("/content/gdrive")

In [3]:
# Libreria pandas para leer el dataset.
df = pd.read_csv('/content/gdrive/MyDrive/SIS 420 INTELIGENCIA ARTIFICAL/datasets/poker-hand-testing.csv')
df

Suit of Card 1  Rank of Card 1  Suit of Card 2  Rank of Card 2  \
0                    1               1               1              13   
1                    3              12               3               2   
2                    1               9               4               6   
3                    1               4               3              13   
4                    3              10               2               7   
...                ...             ...             ...             ...   
999995               3               1               1              12   
999996               3               3               4               5   
999997               1              11               4               7   
999998               3              11               1               8   
999999               2               5               2               9   

        Suit of Card 3  Rank of Card 3  Suit of Card 4  Rank of Card 4  \
0                    2               4               2               3   
1                    3              11               4               5   
2                    1               4               3               2   
3                    2              13               2               1   
4                    1               2               2              11   
...                ...             ...             ...             ...   
999995               2               9               4               9   
999996               2               7               1               4   
999997               3               9               1              13   
999998               1               1               3              13   
999999               4               9               2               3   

        Suit of Card 5  Rank of Card 5  Poker Hand  
0                    1              12           0  
1                    2               5           1  
2                    3               9           1  
3                    3               6           1  
4                    4               9           0  
...                ...             ...         ...  
999995               2               6           1  
999996               4               3           1  
999997               2               7           1  
999998               2               8           1  
999999               3               3           2  

[1000000 rows x 11 columns]

In [ ]:
#eliminamos la primera columan que seria nuestro "id" ya que no nos sirve para el analisis, que esa columna es unnamed
#df = df.drop(df.columns[0], axis=1)

#ahora tenemos un dataframe
#Imprimimos en una tabla el dataset para hacer un analisis mas claro.
df

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000000 entries, 0 to 999999
Data columns (total 11 columns):
 #   Column          Non-Null Count    Dtype
---  ------          --------------    -----
 0   Suit of Card 1  1000000 non-null  int64
 1   Rank of Card 1  1000000 non-null  int64
 2   Suit of Card 2  1000000 non-null  int64
 3   Rank of Card 2  1000000 non-null  int64
 4   Suit of Card 3  1000000 non-null  int64
 5   Rank of Card 3  1000000 non-null  int64
 6   Suit of Card 4  1000000 non-null  int64
 7   Rank of Card 4  1000000 non-null  int64
 8   Suit of Card 5  1000000 non-null  int64
 9   Rank of Card 5  1000000 non-null  int64
 10  Poker Hand      1000000 non-null  int64
dtypes: int64(11)
memory usage: 83.9 MB


In [6]:
#para contar cuantas clases contiene nuestra columna y
class_counts = df["Poker Hand"].value_counts()

#para contar cuantas caracteristicas tiene nuestro dataset, obviamente con sin contar nuestra y, por eso lo dropeamos, tambien dropeamos la primera columna que no tiene nombre
feactures_counts =df.drop(['Poker Hand'], axis=1)
feactures_counts = feactures_counts.shape[1]

print(f"la cantidad de caracteristicas es: { feactures_counts}")
print(f'la cantidad de clases es: 3')

#mostramos la cantidad de clases tiene, y en que cantidad
class_counts

la cantidad de caracteristicas es: 10
la cantidad de clases es: 3


0    501209
1    422498
2     47622
3     21121
4      3885
5      1996
6      1424
7       230
8        12
9         3
Name: Poker Hand, dtype: int64

In [7]:
#almacenamos la cantidad de caracteristicas en una variable
input_layer_size  = feactures_counts;

#almacenamos la cantidad de clases en una variable

#para este caso no es necesario cambiar los valores de la ultima clase, ya que este cuenta con 0,1,2,
# en caso de contener 1,2,3 se tendria que hacer los cambios necesarios para que sea 1,2,0, o simplememte suma 1 a la cantidad de clases.
num_labels = 3;

In [9]:
# hacemos uso del DataFrame llamado 'df' que contiene nuestros datos datos
# y es la columna que contiene las etiquetas de clase (en este caso, 'Credit_Score')

#creamos una variable temporal que contentra toda la columna de 'Credit_Score'
y_temp = df['Poker Hand']

# Para la clase 0

#donde y_temp es igual a 0, separamos los datos en train_class_0, test_class_0
data_class_0 = df[y_temp == 0]
train_class_0, test_class_0 = train_test_split(data_class_0, test_size=0.2, random_state=42)

# Para la clase 1

#donde y_temp es igual a 1, separamos los datos en train_class_1, test_class_1
data_class_1 = df[y_temp == 1]
train_class_1, test_class_1 = train_test_split(data_class_1, test_size=0.2, random_state=42)

# Para la clase 2

#donde y_temp es igual a 2, separamos los datos en train_class_2, test_class_2
data_class_2 = df[y_temp == 2]
train_class_2, test_class_2 = train_test_split(data_class_2, test_size=0.2, random_state=42)

In [ ]:
print(f"Para la clase 0 tenemos una cantidad de: { data_class_0.shape[0]} donde el 80% es: {train_class_0.shape[0]} y el 20% es: {test_class_0.shape[0]}")
print(f"Para la clase 1 tenemos una cantidad de: { data_class_1.shape[0]} donde el 80% es: {train_class_1.shape[0]} y el 20% es: {test_class_1.shape[0]}")
print(f"Para la clase 2 tenemos una cantidad de: { data_class_2.shape[0]} donde el 80% es: {train_class_2.shape[0]} y el 20% es: {test_class_2.shape[0]}")
print(f"La cantidad total de datos es: {df.shape[0]}")

In [11]:
#para la parte de entrenamiento, separamos las caracteristicas de la etiqueta

#para la clase 0
X_train_class_0 = train_class_0.drop(['Poker Hand'], axis=1)
y_train_class_0 = train_class_0['Poker Hand']

#para la clase 1
X_train_class_1 = train_class_1.drop(['Poker Hand'], axis=1)
y_train_class_1 = train_class_1['Poker Hand']

#para la clase 2
X_train_class_2 = train_class_2.drop(['Poker Hand'], axis=1)
y_train_class_2 = train_class_2['Poker Hand']

#ahora para la parte de pruebas, separamos las caracteristicas de la etiqueta
#para la clase 0
X_test_class_0 = test_class_0.drop(['Poker Hand'], axis=1)
y_test_class_0 = test_class_0['Poker Hand']

#para la clase 1
X_test_class_1 = test_class_1.drop(['Poker Hand'], axis=1)
y_test_class_1 = test_class_1['Poker Hand']

#para la clase 2
X_test_class_2 = test_class_2.drop(['Poker Hand'], axis=1)
y_test_class_2 = test_class_2['Poker Hand']

In [12]:
#separando los datos de entrenamiento y pruebas

#para los datos de entrenamiento
X_train = pd.concat([X_train_class_0, X_train_class_1, X_train_class_2]).values
y_train = pd.concat([y_train_class_0, y_train_class_1, y_train_class_2]).values


indices_train = np.random.permutation(len(X_train))
X_train = X_train[indices_train]
y_train = y_train[indices_train]
m_train = len(y_train)

#para los datos de pruebas
X_test = pd.concat([X_test_class_0, X_test_class_1, X_test_class_2]).values
y_test = pd.concat([y_test_class_0, y_test_class_1, y_test_class_2]).values

indices_test = np.random.permutation(len(X_test))
X_test = X_test[indices_test]
y_test = y_test[indices_test]
m_test = len(y_test)


In [14]:
#Visualizamos una cantidad de datos de entrenamiento
print('{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}'.format(
    'X[:,0]', 'X[:, 1]', 'X[:, 2]', 'X[:, 3]', 'X[:, 4]', 'X[:, 5]', 'X[:, 6]', 'X[:, 7]', 'X[:, 8]', 'X[:, 9]','Y'
))
print('-' * 250)

for i in range(10):
    print('{:10.0f}{:10.0f}{:10.0f}{:10.0f}{:10.0f}{:10.0f}{:8.0f}{:10.0f}{:10.0f}{:10.0f}{:10.0f}'.format(
        X_train[i, 0],
        X_train[i, 1],
        X_train[i, 2],
        X_train[i, 3],
        X_train[i, 4],
        X_train[i, 5],
        X_train[i, 6],
        X_train[i, 7],
        X_train[i, 8],
        X_train[i, 9],
        y_train[i]
    ))

print(" ")
print('El 80% de ejemplos para entrenamiento son la cantidad de: {:.0f} de ejemplos'.format( len(X_train)))
print('El 20% de ejemplos para pruebas son la cantidad de: {:.0f} de ejemplos'.format( len(X_test)))
print('La cantidad total de ejemplos es de: {:.0f} de ejemplos'.format( len(df)))

    X[:,0]   X[:, 1]   X[:, 2]   X[:, 3]   X[:, 4]   X[:, 5]   X[:, 6]   X[:, 7]   X[:, 8]   X[:, 9]         Y
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
         1         9         4         1         4        11       2        13         3        13         1
         1         6         3         7         2        13       3         9         1         9         1
         4        10         2         9         4        12       3        11         3        12         1
         3        12         2         9         1         1       4         3         4         1         1
         1        10         2        13         1         9       1        11         3        11         1
         3         1         1        10         2         6       1         8         4     

Funcion de Normalizacion de Caracteristicas

In [15]:
def  featureNormalize(X):
    X_norm = X.copy()

    #creamos un array de ceros con una longitud igual al número de columnas en el array X. La variable mu y sigma se inicializa como este array de ceros.
    mu = np.zeros(X.shape[1])
    sigma = np.zeros(X.shape[1])

    #Creamos el promedio de cada filaa de X
    #media de cada columna
    mu = np.mean(X, axis = 0)

    #desviacion estandar de cada fila de X
    sigma = np.std(X, axis = 0)

    sigma[sigma == 0] = 1

    #normalizamos los datos con la siguiente formula
    X_norm = (X - mu) / sigma

    return X_norm, mu, sigma

Funcion para el Calculo de la Sigmoide

In [16]:
def sigmoid(z):
    z = np.array(z)

    g = np.zeros(z.shape)

    g = 1 / (1 + np.exp(-z))

    return g

Funcion de calculo del costo con Regularizacion

In [17]:
def lrCostFunctionCR(theta, X, y, lambda_):
    #creamos una variable m que contiene la longitud de y
    m = y.size

    # convierte las etiquetas a valores enteros si son boleanos
    if y.dtype == bool:
        y = y.astype(int)

    #inicializamos J y grad
    J = 0
    grad = np.zeros(theta.shape)

    #calculamos h haciendo uso de la funcion sigmoid, donde h es nuestra hipotesis
    h = sigmoid(X.dot(theta.T))

    temp = theta
    temp[0] = 0

    J = (1 / m) * np.sum(-y.dot(np.log(h)) - (1 - y).dot(np.log(1 - h))) + (lambda_ / (2 * m)) * np.sum(np.square(temp))

    grad = (1 / m) * (h - y).dot(X)

    grad = grad + (lambda_ / m) * temp

    return J, grad

Funcion One-vs-all con Regularizacion

In [18]:
def oneVsAllCR(X, y, num_labels, lambda_):
    # algunas variables utiles
    #m es la longitud de y, n es la cantidad de columnas en X
    m, n = X.shape

    # inicializamos la matriz de thetas con ceros, y con n+1 columnas que seria nuestro sesgo
    all_theta = np.zeros((num_labels, n + 1))

    # Agrega unos a la matriz X
    X = np.concatenate([np.ones((m, 1)), X], axis=1)

    # iteramos sobre cada etiqueta (clase) y entrenamos un clasificador
    for c in np.arange(num_labels):
        initial_theta = np.zeros(n + 1)
        options = {'maxiter': 20000}
        res = optimize.minimize(lrCostFunctionCR,
                                initial_theta,
                                (X, (y == c), lambda_),
                                jac=True,
                                method='CG',
                                options=options)

        all_theta[c] = res.x

    return all_theta

Funcion del Calculo del Costo sin Regularizacion

In [19]:
def lrCostFunctionSR(theta, X, y):
    #creamos una variable m que contiene la longitud de y
    m = y.size

    # convierte las etiquetas a valores enteros si son boleanos
    if y.dtype == bool:
        y = y.astype(int)

    #inicializamos J y grad
    J = 0

    # grad = np.zeros(theta.shape)

    #calculamos h haciendo uso de la funcion sigmoid, donde h es nuestra hipotesis
    h = sigmoid(X.dot(theta.T))

    # temp = theta
    # temp[0] = 0

    # J = (1 / m) * np.sum(-y.dot(np.log(h)) - (1 - y).dot(np.log(1 - h))) + (lambda_ / (2 * m)) * np.sum(np.square(temp))
    J = (1 / m) * np.sum(-y.dot(np.log(h)) - (1 - y).dot(np.log(1 - h)))

    grad = (1 / m) * (h - y).dot(X)

    # grad = grad + (lambda_ / m) * temp

    return J, grad

Funcion One-vs-all sin regularizacion

In [20]:
def oneVsAllSR(X, y, num_labels):
    # algunas variables utiles
    #m es la longitud de y, n es la cantidad de columnas en X
    m, n = X.shape

    # inicializamos la matriz de thetas con ceros, y con n+1 columnas que seria nuestro sesgo
    all_theta = np.zeros((num_labels, n + 1))

    # Agrega unos a la matriz X
    X = np.concatenate([np.ones((m, 1)), X], axis=1)

    # iteramos sobre cada etiqueta (clase) y entrenamos un clasificador
    for c in np.arange(num_labels):
        initial_theta = np.zeros(n + 1)
        options = {'maxiter': 20000}
        res = optimize.minimize(lrCostFunctionSR,
                                initial_theta,
                                (X, (y == c)),
                                jac=True,
                                method='CG',
                                options=options)

        all_theta[c] = res.x

    return all_theta

Funcion para Prediccion One-vs-all

In [21]:
def predictOneVsAll(all_theta, X):
    m = X.shape[0];
    num_labels = all_theta.shape[0]

    # inicializamos la matriz de p con ceros
    p = np.zeros(m)

    # añadimos unos a la matriz de X
    X = np.concatenate([np.ones((m, 1)), X], axis=1)
    p = np.argmax(sigmoid(X.dot(all_theta.T)), axis = 1)

    return p

# **Usando 'Regularizacion'**

Clasificacion Multiclase

In [24]:
#estos datos seran usados para el entrenamiento
X_testCR = X_test.copy()
y_testCR = y_test.copy()
m_test_CR = len(y_testCR)

#estos datos seran usados para el test
X_trainCR = X_train.copy()
y_trainCR = y_train.copy()
m_train_CR = len(y_trainCR)

#Visualizamos una cantidad de datos de entrenamiento
print('{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}'.format(
    'X[:,0]', 'X[:, 1]', 'X[:, 2]', 'X[:, 3]', 'X[:, 4]', 'X[:, 5]', 'X[:, 6]', 'X[:, 7]', 'X[:, 8]', 'X[:, 9]','Y'
))
print('-' * 250)

for i in range(10):
    print('{:10.0f}{:10.0f}{:10.0f}{:10.0f}{:10.0f}{:10.0f}{:8.0f}{:10.0f}{:10.0f}{:10.0f}{:10.0f}'.format(
        X_trainCR[i, 0],
        X_trainCR[i, 1],
        X_trainCR[i, 2],
        X_trainCR[i, 3],
        X_trainCR[i, 4],
        X_trainCR[i, 5],
        X_trainCR[i, 6],
        X_trainCR[i, 7],
        X_trainCR[i, 8],
        X_trainCR[i, 9],
        y_trainCR[i]
    ))
print(" ")
print('El 80% de ejemplos para entrenamiento son la cantidad de: {:.0f} de ejemplos'.format( len(X_train)))
print('El 20% de ejemplos para pruebas son la cantidad de: {:.0f} de ejemplos'.format( len(X_test)))
print('La cantidad total de ejemplos es de: {:.0f} de ejemplos'.format( len(df)))

    X[:,0]   X[:, 1]   X[:, 2]   X[:, 3]   X[:, 4]   X[:, 5]   X[:, 6]   X[:, 7]   X[:, 8]   X[:, 9]         Y
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
         1         9         4         1         4        11       2        13         3        13         1
         1         6         3         7         2        13       3         9         1         9         1
         4        10         2         9         4        12       3        11         3        12         1
         3        12         2         9         1         1       4         3         4         1         1
         1        10         2        13         1         9       1        11         3        11         1
         3         1         1        10         2         6       1         8         4     

Normalizacion de Caracteristicas

In [26]:
X_norm_CR, mu_CR, sigma_CR = featureNormalize(X_trainCR)

# imprimir todos las X_norm de datos solo 10
print('{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}'.format(
    'X[:,0]', 'X[:, 1]', 'X[:, 2]', 'X[:, 3]', 'X[:, 4]', 'X[:, 5]', 'X[:, 6]', 'X[:, 7]', 'X[:, 8]', 'X[:, 9]','Y'
))
print('-' * 250)

for i in range(10):
    print('{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}'.format(
        X_norm_CR[i, 0],
        X_norm_CR[i, 1],
        X_norm_CR[i, 2],
        X_norm_CR[i, 3],
        X_norm_CR[i, 4],
        X_norm_CR[i, 5],
        X_norm_CR[i, 6],
        X_norm_CR[i, 7],
        X_norm_CR[i, 8],
        X_norm_CR[i, 9],
        y_trainCR[i]
    ))

    X[:,0]   X[:, 1]   X[:, 2]   X[:, 3]   X[:, 4]   X[:, 5]   X[:, 6]   X[:, 7]   X[:, 8]   X[:, 9]         Y
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
     -1.34      0.54      1.34     -1.60      1.34      1.07     -0.45      1.60      0.45      1.61      1.00
     -1.34     -0.27      0.45     -0.00     -0.45      1.60      0.45      0.53     -1.34      0.54      1.00
      1.34      0.80     -0.45      0.53      1.34      1.34      0.45      1.07      0.45      1.34      1.00
      0.45      1.34     -0.45      0.53     -1.34     -1.60      1.34     -1.07      1.34     -1.60      1.00
     -1.34      0.80     -0.45      1.60     -1.34      0.53     -1.34      1.07      0.45      1.07      1.00
      0.45     -1.60     -1.34      0.80     -0.45     -0.27     -1.34      0.27   

Clasificacion One-vs-all

In [27]:
lambda_ = 1000
all_theta_CR = oneVsAllCR(X_norm_CR, y_trainCR, num_labels, lambda_)
print(all_theta_CR.shape)

#imprimimos todos los thetas para cada clase
for i in range(all_theta_CR.shape[1]):
    print("Thetha",i," para la clase 0:", all_theta_CR[0,i]," ; ", "Thetha",i," para la clase 1:", all_theta_CR[1,i]," ; ", "Thetha",i," para la clase 2:", all_theta_CR[2,i])

(3, 11)
Thetha 0  para la clase 0: 0.06403891084625535  ;  Thetha 0  para la clase 1: -0.2616188400925089  ;  Thetha 0  para la clase 2: -2.965287671820982
Thetha 1  para la clase 0: 0.0011258435633153333  ;  Thetha 1  para la clase 1: -0.00024820101143669814  ;  Thetha 1  para la clase 2: -0.004628056860695274
Thetha 2  para la clase 0: -0.003897835061735982  ;  Thetha 2  para la clase 1: 0.004566558533580316  ;  Thetha 2  para la clase 2: -0.0031480261318697156
Thetha 3  para la clase 0: 0.0005204319188535908  ;  Thetha 3  para la clase 1: -0.0010089832827025428  ;  Thetha 3  para la clase 2: 0.0025184768790122236
Thetha 4  para la clase 0: 0.0008539066779424824  ;  Thetha 4  para la clase 1: 0.00020214427912251085  ;  Thetha 4  para la clase 2: -0.005625411142937098
Thetha 5  para la clase 0: 0.0009529036935747852  ;  Thetha 5  para la clase 1: -0.0011327255695461188  ;  Thetha 5  para la clase 2: 0.0008773914249843321
Thetha 6  para la clase 0: -0.0014977869458178013  ;  Thetha 6  

Validaciones

In [28]:
# print(X_test.shape)
pred_test_CR = predictOneVsAll(all_theta_CR, X_norm_CR)
print('Precision del conjunto de entrenamiento: {:.5f}%'.format(np.mean(pred_test_CR == y_trainCR) * 100))

Precision del conjunto de entrenamiento: 51.60039%


In [29]:
X_test1_CR = X_testCR.copy()
# print(X_test.shape)
X_test1_CR = (X_test1_CR - mu_CR) / sigma_CR
pred_train_CR = predictOneVsAll(all_theta_CR, X_test1_CR)
print('Precision del conjunto de entrenamiento: {:.5f}%'.format(np.mean(pred_train_CR == y_testCR) * 100))

Precision del conjunto de entrenamiento: 51.60012%


In [30]:
#usando el mu y sigma calculado anteriormente, para realizar la normalizacion de los datos de prueba
X_test1_CR = (X_testCR.copy() - mu_CR) / sigma_CR
X_test1_CR = np.concatenate([np.ones((len(X_test1_CR), 1)), X_test1_CR], axis=1)

print(X_test1_CR.shape)

#para este ejemplo, vamos a mostrar las predicciones de las primeras 30 predicciones y compararlas con las reales
p = np.argmax(sigmoid(X_test1_CR.dot(all_theta_CR.T)), axis = 1)

for i in range(30):
    print(f"Prediccion: {p[i]}, Real: {y_test[i]}")

(194267, 11)
Prediccion: 0, Real: 1
Prediccion: 0, Real: 0
Prediccion: 0, Real: 1
Prediccion: 0, Real: 0
Prediccion: 0, Real: 1
Prediccion: 0, Real: 0
Prediccion: 0, Real: 1
Prediccion: 0, Real: 1
Prediccion: 0, Real: 1
Prediccion: 0, Real: 0
Prediccion: 0, Real: 1
Prediccion: 0, Real: 2
Prediccion: 0, Real: 2
Prediccion: 0, Real: 1
Prediccion: 0, Real: 1
Prediccion: 0, Real: 0
Prediccion: 0, Real: 0
Prediccion: 0, Real: 0
Prediccion: 0, Real: 0
Prediccion: 0, Real: 0
Prediccion: 0, Real: 0
Prediccion: 0, Real: 1
Prediccion: 0, Real: 0
Prediccion: 0, Real: 0
Prediccion: 0, Real: 1
Prediccion: 0, Real: 1
Prediccion: 0, Real: 0
Prediccion: 0, Real: 1
Prediccion: 0, Real: 1
Prediccion: 0, Real: 1


# **Usando sin 'Regularizacion'**

Clasificacion Multiclase

In [31]:
#estos datos seran usados para el entrenamiento
X_testSR = X_test.copy()
y_testSR = y_test.copy()
m_test_SR = len(y_testSR)

#estos datos seran usados para el test
X_trainSR = X_train.copy()
y_trainSR = y_train.copy()
m_train_SR = len(y_trainSR)

#Visualizamos una cantidad de datos de entrenamiento
print('{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}'.format(
    'X[:,0]', 'X[:, 1]', 'X[:, 2]', 'X[:, 3]', 'X[:, 4]', 'X[:, 5]', 'X[:, 6]', 'X[:, 7]', 'X[:, 8]', 'X[:, 9]','Y'
))
print('-' * 250)

for i in range(10):
    print('{:10.0f}{:10.0f}{:10.0f}{:10.0f}{:10.0f}{:10.0f}{:8.0f}{:10.0f}{:10.0f}{:10.0f}{:10.0f}'.format(
        X_trainSR[i, 0],
        X_trainSR[i, 1],
        X_trainSR[i, 2],
        X_trainSR[i, 3],
        X_trainSR[i, 4],
        X_trainSR[i, 5],
        X_trainSR[i, 6],
        X_trainSR[i, 7],
        X_trainSR[i, 8],
        X_trainSR[i, 9],
        y_trainSR[i]
    ))

print(" ")
print('El 80% de ejemplos para entrenamiento son la cantidad de: {:.0f} de ejemplos'.format( len(X_train)))
print('El 20% de ejemplos para pruebas son la cantidad de: {:.0f} de ejemplos'.format( len(X_test)))
print('La cantidad total de ejemplos es de: {:.0f} de ejemplos'.format( len(df)))

    X[:,0]   X[:, 1]   X[:, 2]   X[:, 3]   X[:, 4]   X[:, 5]   X[:, 6]   X[:, 7]   X[:, 8]   X[:, 9]         Y
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
         1         9         4         1         4        11       2        13         3        13         1
         1         6         3         7         2        13       3         9         1         9         1
         4        10         2         9         4        12       3        11         3        12         1
         3        12         2         9         1         1       4         3         4         1         1
         1        10         2        13         1         9       1        11         3        11         1
         3         1         1        10         2         6       1         8         4     

Normalizacion de Caracteristicas

In [32]:
X_norm_SR, mu_SR, sigma_SR = featureNormalize(X_trainSR)

# imprimir todos las X_norm de datos solo 10
print('{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}'.format(
    'X[:,0]', 'X[:, 1]', 'X[:, 2]', 'X[:, 3]', 'X[:, 4]', 'X[:, 5]', 'X[:, 6]', 'X[:, 7]', 'X[:, 8]', 'X[:, 9]','Y'
))
print('-' * 250)

for i in range(10):
    print('{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}'.format(
        X_norm_SR[i, 0],
        X_norm_SR[i, 1],
        X_norm_SR[i, 2],
        X_norm_SR[i, 3],
        X_norm_SR[i, 4],
        X_norm_SR[i, 5],
        X_norm_SR[i, 6],
        X_norm_SR[i, 7],
        X_norm_SR[i, 8],
        X_norm_SR[i, 9],
        y_trainSR[i]
    ))

    X[:,0]   X[:, 1]   X[:, 2]   X[:, 3]   X[:, 4]   X[:, 5]   X[:, 6]   X[:, 7]   X[:, 8]   X[:, 9]         Y
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
     -1.34      0.54      1.34     -1.60      1.34      1.07     -0.45      1.60      0.45      1.61      1.00
     -1.34     -0.27      0.45     -0.00     -0.45      1.60      0.45      0.53     -1.34      0.54      1.00
      1.34      0.80     -0.45      0.53      1.34      1.34      0.45      1.07      0.45      1.34      1.00
      0.45      1.34     -0.45      0.53     -1.34     -1.60      1.34     -1.07      1.34     -1.60      1.00
     -1.34      0.80     -0.45      1.60     -1.34      0.53     -1.34      1.07      0.45      1.07      1.00
      0.45     -1.60     -1.34      0.80     -0.45     -0.27     -1.34      0.27   

Clasificacion One-vs-all

In [33]:
all_theta_SR = oneVsAllSR(X_norm_SR, y_trainSR, num_labels)
print(all_theta_CR.shape)

#imprimimos todos los thetas para cada clase
for i in range(all_theta_SR.shape[1]):
    print("Thetha",i," para la clase 0:", all_theta_SR[0,i]," ; ", "Thetha",i," para la clase 1:", all_theta_SR[1,i]," ; ", "Thetha",i," para la clase 2:", all_theta_SR[2,i])

(3, 11)
Thetha 0  para la clase 0: 0.06403895506661354  ;  Thetha 0  para la clase 1: -0.26162066804660145  ;  Thetha 0  para la clase 2: -2.965290521510949
Thetha 1  para la clase 0: 0.0011327072269034606  ;  Thetha 1  para la clase 1: -0.0002493296590928227  ;  Thetha 1  para la clase 2: -0.004755457564471026
Thetha 2  para la clase 0: -0.003919701878991151  ;  Thetha 2  para la clase 1: 0.004590514904348821  ;  Thetha 2  para la clase 2: -0.003235973148316631
Thetha 3  para la clase 0: 0.0005241547799975284  ;  Thetha 3  para la clase 1: -0.0010141725340561574  ;  Thetha 3  para la clase 2: 0.002583478195115252
Thetha 4  para la clase 0: 0.0008565947620617298  ;  Thetha 4  para la clase 1: 0.00020261615610023617  ;  Thetha 4  para la clase 2: -0.005779217035263437
Thetha 5  para la clase 0: 0.0009589634903982631  ;  Thetha 5  para la clase 1: -0.001138553188764356  ;  Thetha 5  para la clase 2: 0.0008983977210261451
Thetha 6  para la clase 0: -0.0015071961426174954  ;  Thetha 6  par

Validaciones

In [34]:
# print(X_test.shape)
pred_test_SR = predictOneVsAll(all_theta_SR, X_norm_SR)
print('Precision del conjunto de entrenamiento: {:.5f}%'.format(np.mean(pred_test_SR == y_trainSR) * 100))

Precision del conjunto de entrenamiento: 51.60039%


In [35]:
X_test1_SR = X_testSR.copy()
# print(X_test.shape)
X_test1_SR = (X_test1_SR - mu_SR) / sigma_SR
pred_train_SR = predictOneVsAll(all_theta_SR, X_test1_SR)
print('Precision del conjunto de entrenamiento: {:.5f}%'.format(np.mean(pred_train_SR == y_testSR) * 100))

Precision del conjunto de entrenamiento: 51.60012%


In [36]:
#usando el mu y sigma calculado anteriormente, para realizar la normalizacion de los datos de prueba
X_test1_SR = (X_testSR.copy() - mu_SR) / sigma_SR
X_test1_SR = np.concatenate([np.ones((len(X_test1_SR), 1)), X_test1_SR], axis=1)

print(X_test1_SR.shape)

#para este ejemplo, vamos a mostrar las predicciones de las primeras 30 predicciones y compararlas con las reales
p = np.argmax(sigmoid(X_test1_SR.dot(all_theta_SR.T)), axis = 1)

for i in range(30):
    print(f"Prediccion: {p[i]}, Real: {y_test[i]}")

(194267, 11)
Prediccion: 0, Real: 1
Prediccion: 0, Real: 0
Prediccion: 0, Real: 1
Prediccion: 0, Real: 0
Prediccion: 0, Real: 1
Prediccion: 0, Real: 0
Prediccion: 0, Real: 1
Prediccion: 0, Real: 1
Prediccion: 0, Real: 1
Prediccion: 0, Real: 0
Prediccion: 0, Real: 1
Prediccion: 0, Real: 2
Prediccion: 0, Real: 2
Prediccion: 0, Real: 1
Prediccion: 0, Real: 1
Prediccion: 0, Real: 0
Prediccion: 0, Real: 0
Prediccion: 0, Real: 0
Prediccion: 0, Real: 0
Prediccion: 0, Real: 0
Prediccion: 0, Real: 0
Prediccion: 0, Real: 1
Prediccion: 0, Real: 0
Prediccion: 0, Real: 0
Prediccion: 0, Real: 1
Prediccion: 0, Real: 1
Prediccion: 0, Real: 0
Prediccion: 0, Real: 1
Prediccion: 0, Real: 1
Prediccion: 0, Real: 1


# **Informe**

1. Respecto a la presicion del modelo


*   Presicion con regularizacion 51.60012%
*   Presicion sin regularizacion 51.60012%

Ambas son similiares, por lo que no se puede dar una conclusion coherente. Sim embargo, cabe resaltar, que la regularizacion proporciona una ventaja adicional al ayudar a prevenir el sobreajuste, actua como una medida preventiva para garantizar que el modelo generalice bien a datos no visto.

